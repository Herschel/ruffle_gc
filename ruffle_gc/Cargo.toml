[package]
name = "ruffle_gc"
version = "0.1.0"
authors = ["Mike Welsh <mwelsh@gmail.com>"]
edition = "2021"

[dependencies]
bitflags = "1.2"
generational-arena = "0.2.8"
once_cell = "1.7"
ruffle_gc_derive = { path = "../ruffle_gc_derive" }

[dev-dependencies]
trybuild = "1.0"
