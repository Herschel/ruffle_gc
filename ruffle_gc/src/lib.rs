mod context;
mod gc;
mod lifetime;
mod root;
mod trace;
mod weak;

pub use context::GcContext;
pub use gc::{Gc, GcVtbl};
pub use lifetime::GcLifetime;
pub use root::{GcHeapRoot, GcRoot, GcRootData};
pub use trace::Trace;
pub use weak::GcWeak;

pub use ruffle_gc_derive::Gc;

pub(crate) use gc::{GcData, GcDataPtr, GcFlags};
pub(crate) use weak::WeakId;

/// Creates a new GC root on the stack.
#[macro_export]
macro_rules! pin_root {
    ($name:ident $(,)?) => {
        let mut $name = $name;
        let mut $name = unsafe { ruffle_gc::GcRoot::new($name) };
        #[allow(unused_mut)]
        let mut $name = $name.pin();
    };
}
