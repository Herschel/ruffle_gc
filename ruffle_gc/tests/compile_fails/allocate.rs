use ruffle_gc::GcContext;

fn main() {
    let mut ctx = GcContext::new().unwrap();
    // Mutably borrows context so that garbage collection can't occur until the GC data is stored or rooted:
    let data = ctx.allocate("Test".to_string());
    // Shouldn't be able to collect while an unrooted borrow exists:
    ctx.collect(); // Can't mutably borrow context twice.

    println!("{}", data.borrow(&ctx));
}
