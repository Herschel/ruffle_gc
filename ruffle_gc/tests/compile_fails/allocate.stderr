error[E0499]: cannot borrow `ctx` as mutable more than once at a time
  --> tests/compile_fails/allocate.rs:8:5
   |
6  |     let data = ctx.allocate("Test".to_string());
   |                -------------------------------- first mutable borrow occurs here
7  |     // Shouldn't be able to collect while an unrooted borrow exists:
8  |     ctx.collect(); // Can't mutably borrow context twice.
   |     ^^^^^^^^^^^^^ second mutable borrow occurs here
9  |
10 |     println!("{}", data.borrow(&ctx));
   |                    ---- first borrow later used here

error[E0502]: cannot borrow `ctx` as immutable because it is also borrowed as mutable
  --> tests/compile_fails/allocate.rs:10:32
   |
6  |     let data = ctx.allocate("Test".to_string());
   |                -------------------------------- mutable borrow occurs here
...
10 |     println!("{}", data.borrow(&ctx));
   |                         ------ ^^^^ immutable borrow occurs here
   |                         |
   |                         mutable borrow later used by call
